tianshou/trainer/base.py

```diff
@@ -0,0 +1,452 @@
+
+        self.train_fn = train_fn
+        self.test_fn = test_fn
+        self.stop_fn = stop_fn
+        self.save_fn = save_fn
+        self.save_checkpoint_fn = save_checkpoint_fn
+
+        self.reward_metric = reward_metric
+        self.verbose = verbose
+        self.test_in_train = test_in_train
+        self.resume_from_log = resume_from_log
+
+        self.is_run = False
+        self.last_rew, self.last_len = 0.0, 0
+
+        self.epoch = self.start_epoch
+        self.best_epoch = self.start_epoch
+        self.stop_fn_flag = False
+        self.iter_num = 0
+
+        self.policy_update_fn = getattr(self, f"{learning_type}_update")
+
+    def reset(self) -> None:
+        """Initialize or reset the instance to yield a new iterator from zero."""
+        self.is_run = False
+        self.env_step = 0
+        if self.resume_from_log:
+            self.start_epoch, self.env_step, self.gradient_step = \
+                self.logger.restore_data()
+
+        self.last_rew, self.last_len = 0.0, 0
+        self.start_time = time.time()
+        if self.train_collector is not None:
+            self.train_collector.reset_stat()
+
+            if self.train_collector.policy != self.policy:
+                self.test_in_train = False
+            elif self.test_collector is None:
+                self.test_in_train = False
+
+        if self.test_collector is not None:
+            assert self.episode_per_test is not None
+            self.test_collector.reset_stat()
+            test_result = test_episode(
+                self.policy, self.test_collector, self.test_fn, self.start_epoch,
+                self.episode_per_test, self.logger, self.env_step, self.reward_metric
+            )
+            self.best_epoch = self.start_epoch
+            self.best_reward, self.best_reward_std = \
+                test_result["rew"], test_result["rew_std"]
+        if self.save_fn:
+            self.save_fn(self.policy)
+
+        self.epoch = self.start_epoch
+        self.stop_fn_flag = False
+        self.iter_num = 0
+
+    def __iter__(self):  # type: ignore
+        self.reset()
+        return self
+
+    def __next__(self) -> Union[None, Tuple[int, Dict[str, Any], Dict[str, Any]]]:
+        """Perform one epoch (both train and eval)."""
+        self.epoch += 1
+        self.iter_num += 1
+
+        if self.iter_num > 1:
+
+            # iterator exhaustion check
+            if self.epoch >= self.max_epoch:
+                if self.test_collector is None and self.save_fn:
+                    self.save_fn(self.policy)
+                raise StopIteration
+
+            # exit flag 1, when stop_fn succeeds in train_step or test_step
+            if self.stop_fn_flag:
+                raise StopIteration
+
+            # stop_fn criterion
+            if self.test_collector is not None and self.stop_fn and self.stop_fn(
+                self.best_reward
+            ):
+                raise StopIteration
```

Comment:
Is there any condition that can hit this line? https://app.codecov.io/gh/thu-ml/tianshou/compare/559/tree/tianshou/trainer/base.py#D1L256